Support `mtl-2.3-rc4`

Dependencies breaking with `mtl-2.3`:
- [x] `equivalence` https://github.com/pa-ba/equivalence/issues/14 
- [ ] `exceptions` https://github.com/ekmett/exceptions/issues/82
- [x] `constraints` https://github.com/ekmett/constraints/issues/110 (Fixed by `constraints-0.13.3`)

Problems in our code:
- [x] `transformers-0.6`, enabled by `mtl-2.3` breaks compilation of `instance MonadTrans TCMT`.

# Context: 

Release of `mtl-2.3` is imminent, see
- https://github.com/haskell/mtl/releases/tag/v2.3-rc4
- https://mail.haskell.org/pipermail/libraries/2021-November/031523.html (quoted below)

## Announcement mail by Emily Pillmore

Chessai and I are excited to announce a release candidate for `mtl`: [`mtl-2.3-rc3`]( https://github.com/haskell/mtl/releases/tag/v2.3-rc3 )!

### Timeline

The timeline for release will be as follows:

- This announcement marks the start of the timeline

- We will give 2 weeks of testing before considering release

- If no major issues are filed before then, `mtl-2.3-rc3` will be released as `mtl-2.3`, and if issues are found, they will be amended, and a new release candidate will be announced, resetting the 2 week period.

To test `mtl-2.3-rc3` for yourself, please feel free to add the following to your `cabal.project`:

```
source-repository-package
  type: git
  location: https://github.com/haskell/mtl.git
  tag: 5d0f62b8007bb96e49f36a5544741cfe96a97130
```

or, if you're a stack user add this entry to your `extra-deps`:

```
- git: https://github.com/haskell/mtl.git
  commit: 5d0f62b8007bb96e49f36a5544741cfe96a97130
```

And make sure to adjust all bounds/`allow-newer` accordingly. Please note that this release of `mtl` is a full major version release, and *will be the last of the 2.x series before work begins on updating the mtl class hierarchy*.

### Changelog

* Add instances for `Control.Monad.Trans.Writer.CPS` and `Control.Monad.Trans.RWS.CPS` from `transformers` 0.5.6 and add `Control.Monad.Writer.CPS` and `Control.Monad.RWS.CPS`.

* `Control.Monad.Cont` now re-exports `evalCont` and `evalContT`

* Add `tryError`, `withError`, `handleError`, and `mapError` to

`Control.Monad.Error.Class`, and re-export from `Control.Monad.Except`.

* Remove `Control.Monad.List` and `Control.Monad.Error`

* Remove instances of deprecated `ListT` and `ErrorT`

* Remove re-exports of `Error`

* Add instances for `Control.Monad.Trans.Accum` and ` Control.Monad.Trans.Select ( http://control.monad.trans.select/ ) `

* Remove re-exports of `Control.Monad`, `Control.Monad.Fix` and `Data.Monoid` modules

I'd like to thank the many contributors who offered patches, tickets, and other help in the preparation of this release. We appreciate all of your help!

Happy hacking!

Emily

Seems we are hit by the following innovation in `transformers-0.6`:
> 0.6.0.0 Ross Paterson <R.Paterson@city.ac.uk> Jul 2021
> * Added quantified constraint to MonadTrans (for GHC >= 8.6)
```haskell
class (forall m. Monad m => Monad (t m)) => MonadTrans t where
```
Problem here:
https://github.com/agda/agda/blob/c7499fb8a85eb8be160246ad5a2f0996c43eea9f/src/full/Agda/TypeChecking/Monad/Base.hs#L4263-L4276
```
src/full/Agda/TypeChecking/Monad/Base.hs:4263:10: error:
    • Could not deduce (MonadIO m)
        arising from the superclasses of an instance declaration
      from the context: Monad m
        bound by a quantified context
        at src/full/Agda/TypeChecking/Monad/Base.hs:1:1
      Possible fix:
        add (MonadIO m) to the context of a quantified context
    • In the instance declaration for ‘MonadTrans TCMT’
```

Reported this problem upstream: 
- https://hub.darcs.net/ross/transformers/issue/85
- https://gitlab.haskell.org/ghc/ghc/-/issues/19922#note_405825

--- a/Agda.cabal
+++ b/Agda.cabal
@@ -240,11 +240,9 @@ library
                 , haskeline >= 0.7.2.3 && < 0.9
                 -- monad-control-1.0.1.0 is the first to contain liftThrough
                 , monad-control >= 1.0.1.0 && < 1.1
-                -- mtl-2.1 contains a severe bug.
-                --
                 -- mtl >= 2.2 && < 2.2.1 doesn't export Control.Monad.Except.
                 -- Need mtl 2.2.2  for export of Control.Monad.IdentityT (ghc 8.2.2+)
-                , mtl >= 2.2.1 && < 2.3
+                , mtl >= 2.2.1 && < 2.4
                 , murmur-hash >= 0.1 && < 0.2
                 , parallel >= 3.2.2.0 && < 3.3
                 , pretty >= 1.1.3.3 && < 1.2
@@ -260,6 +258,12 @@ library
                 , unordered-containers >= 0.2.5.0 && < 0.3
                 , uri-encode >= 1.5.0.4 && < 1.6
                 , zlib == 0.6.*
+
+  -- Andreas, 2022-02-02, issue #5659:
+  -- Build failure with transformers-0.6.0.{0,1,2} and GHC 8.6.
+  -- Transformers-0.6.0.3 might restored GHC 8.6 buildability.
+  if impl(ghc == 8.6.*)
+    build-depends: transformers < 0.6 || >= 0.6.0.3
 
   -- Andreas, 2021-03-10:
   -- All packages we depend upon should be mentioned in an unconditional

--- a/src/full/Agda/TypeChecking/Monad/Base.hs
+++ b/src/full/Agda/TypeChecking/Monad/Base.hs
@@ -4233,31 +4233,31 @@ pureTCM f = TCM $ \ r e -> do
 -- [1] When compiled with -auto-all and run with -p: roughly 750%
 -- faster for one example.
 
-returnTCMT :: MonadIO m => a -> TCMT m a
-returnTCMT = \x -> TCM $ \_ _ -> return x
+returnTCMT :: Applicative m => a -> TCMT m a
+returnTCMT = \x -> TCM $ \_ _ -> pure x
 {-# INLINE returnTCMT #-}
 
-bindTCMT :: MonadIO m => TCMT m a -> (a -> TCMT m b) -> TCMT m b
+bindTCMT :: Monad m => TCMT m a -> (a -> TCMT m b) -> TCMT m b
 bindTCMT = \(TCM m) k -> TCM $ \r e -> m r e >>= \x -> unTCM (k x) r e
 {-# INLINE bindTCMT #-}
 
-thenTCMT :: MonadIO m => TCMT m a -> TCMT m b -> TCMT m b
-thenTCMT = \(TCM m1) (TCM m2) -> TCM $ \r e -> m1 r e >> m2 r e
+thenTCMT :: Applicative m => TCMT m a -> TCMT m b -> TCMT m b
+thenTCMT = \(TCM m1) (TCM m2) -> TCM $ \r e -> m1 r e *> m2 r e
 {-# INLINE thenTCMT #-}
 
-instance MonadIO m => Functor (TCMT m) where
+instance Functor m => Functor (TCMT m) where
   fmap = fmapTCMT
 
-fmapTCMT :: MonadIO m => (a -> b) -> TCMT m a -> TCMT m b
+fmapTCMT :: Functor m => (a -> b) -> TCMT m a -> TCMT m b
 fmapTCMT = \f (TCM m) -> TCM $ \r e -> fmap f (m r e)
 {-# INLINE fmapTCMT #-}
 
-instance MonadIO m => Applicative (TCMT m) where
+instance Applicative m => Applicative (TCMT m) where
   pure  = returnTCMT
   (<*>) = apTCMT
 
-apTCMT :: MonadIO m => TCMT m (a -> b) -> TCMT m a -> TCMT m b
-apTCMT = \(TCM mf) (TCM m) -> TCM $ \r e -> ap (mf r e) (m r e)
+apTCMT :: Applicative m => TCMT m (a -> b) -> TCMT m a -> TCMT m b
+apTCMT = \(TCM mf) (TCM m) -> TCM $ \r e -> mf r e <*> m r e
 {-# INLINE apTCMT #-}
 
 instance MonadTrans TCMT where

--- a/src/full/Agda/TypeChecking/Monad/Base.hs
+++ b/src/full/Agda/TypeChecking/Monad/Base.hs
@@ -4264,7 +4264,14 @@ instance MonadTrans TCMT where
     lift m = TCM $ \_ _ -> m
 
 -- We want a special monad implementation of fail.
+#if __GLASGOW_HASKELL__ < 808
 instance MonadIO m => Monad (TCMT m) where
+#else
+-- Andreas, 2022-02-02, issue #5659:
+-- @transformers-0.6@ requires exactly a @Monad@ superclass constraint here
+-- if we want @instance MonadTrans TCMT@.
+instance Monad m => Monad (TCMT m) where
+#endif
     return = pure
     (>>=)  = bindTCMT
     (>>)   = (*>)

--- a/src/full/Agda/TypeChecking/Monad/Base.hs-boot
+++ b/src/full/Agda/TypeChecking/Monad/Base.hs-boot
@@ -26,8 +26,8 @@ data TCEnv
 data TCState
 newtype TCMT m a = TCM { unTCM :: IORef TCState -> TCEnv -> m a }
 
-instance MonadIO m => Applicative (TCMT m)
-instance MonadIO m => Functor (TCMT m)
+instance Applicative m => Applicative (TCMT m)
+instance Functor m => Functor (TCMT m)
 instance MonadIO m => Monad (TCMT m)
 instance MonadIO m => MonadIO (TCMT m)
 

--- a/src/full/Agda/TypeChecking/Monad/Base.hs-boot
+++ b/src/full/Agda/TypeChecking/Monad/Base.hs-boot
@@ -1,3 +1,5 @@
+{-# LANGUAGE CPP #-}
+
 module Agda.TypeChecking.Monad.Base where
 
 import Control.Monad.IO.Class (MonadIO)
@@ -28,8 +30,16 @@ newtype TCMT m a = TCM { unTCM :: IORef TCState -> TCEnv -> m a }
 
 instance Applicative m => Applicative (TCMT m)
 instance Functor m => Functor (TCMT m)
-instance MonadIO m => Monad (TCMT m)
 instance MonadIO m => MonadIO (TCMT m)
+
+#if __GLASGOW_HASKELL__ < 808
+instance MonadIO m => Monad (TCMT m) where
+#else
+-- Andreas, 2022-02-02, issue #5659:
+-- @transformers-0.6@ requires exactly a @Monad@ superclass constraint here
+-- if we want @instance MonadTrans TCMT@.
+instance Monad m => Monad (TCMT m) where
+#endif
 
 type TCM = TCMT IO
 

